File: _submissions/18UEE071_Ayush/Day3/insertionSortList.cpp
ListNode* Solution::insertionSortList(ListNode* A) 
{
    ListNode* a=A;
    ListNode* b=A;
    
    for(a=a;a;a=a->next)
    {
        for(b=a->next;b;b=b->next )
        {
            if(a->val >= b->val )
            {
                int temp = a->val;
                a->val = b->val;
                b->val = temp;
            }
        }
    }
    return A;
}

File: _submissions/19UCS159_Tharun/Day3/insertionSortList.java
class ListNode {
  public int val;
  public ListNode next;

  ListNode(int x) {
    val = x;
    next = null;
  }
}

public class Solution {
  public ListNode insertionSortList(ListNode A) {
    if (A == null)
      return null;
    ListNode curr = A.next;
    ListNode root = A;
    root.next = null;
    while (curr != null) {
      ListNode temp = curr.next;
      root = insert(curr, root);
      curr = temp;
    }
    return root;
  }

  private ListNode insert(ListNode curr, ListNode root) {
    ListNode temp = root;
    if (root.val > curr.val) {
      root = curr;
      root.next = temp;
      return root;
    }
    while (temp.next != null && temp.next.val <= curr.val) {
      temp = temp.next;
    }
    ListNode nextNode = temp.next;
    temp.next = curr;
    curr.next = nextNode;
    return root;
  }
}
